substrate/frame/delegated-staking/src/types.rs

```diff
@@ -0,0 +1,308 @@
+			.defensive_ok_or(ArithmeticError::Overflow)?;
+
+		Ok(Agent {
+			ledger: AgentLedger {
+				total_delegated: new_total_delegated,
+				unclaimed_withdrawals: new_unclaimed_withdrawals,
+				..self.ledger
+			},
+			..self
+		})
+	}
+
+	/// Add funds that are withdrawn from [Config::CoreStaking] to be claimed by delegators later.
+	pub(crate) fn add_unclaimed_withdraw(
+		self,
+		amount: BalanceOf<T>,
+	) -> Result<Self, DispatchError> {
+		let new_unclaimed_withdrawals = self
+			.ledger
+			.unclaimed_withdrawals
+			.checked_add(&amount)
+			.defensive_ok_or(ArithmeticError::Overflow)?;
+
+		Ok(Agent {
+			ledger: AgentLedger { unclaimed_withdrawals: new_unclaimed_withdrawals, ..self.ledger },
+			..self
+		})
+	}
+
+	/// Amount that is delegated but not bonded yet.
+	///
+	/// This importantly does not include `unclaimed_withdrawals` as those should not be bonded
+	/// again unless explicitly requested.
+	pub(crate) fn available_to_bond(&self) -> BalanceOf<T> {
+		let bonded_stake = self.bonded_stake();
+		let stakeable = self.ledger.stakeable_balance();
+
+		defensive_assert!(
+			stakeable >= bonded_stake,
+			"cannot be bonded with more than total amount delegated to agent"
+		);
+
+		stakeable.saturating_sub(bonded_stake)
+	}
+
+	/// Remove slashes from the `AgentLedger`.
+	pub(crate) fn remove_slash(self, amount: BalanceOf<T>) -> Self {
+		let pending_slash = self.ledger.pending_slash.defensive_saturating_sub(amount);
+		let total_delegated = self.ledger.total_delegated.defensive_saturating_sub(amount);
+
+		Agent { ledger: AgentLedger { pending_slash, total_delegated, ..self.ledger }, ..self }
+	}
+
+	/// Get the total stake of agent bonded in [`Config::CoreStaking`].
+	pub(crate) fn bonded_stake(&self) -> BalanceOf<T> {
+		T::CoreStaking::total_stake(&self.key).unwrap_or(Zero::zero())
+	}
+
+	/// Returns true if the agent is bonded in [`Config::CoreStaking`].
+	pub(crate) fn is_bonded(&self) -> bool {
+		T::CoreStaking::stake(&self.key).is_ok()
+	}
+
+	/// Returns the reward account registered by the agent.
+	pub(crate) fn reward_account(&self) -> &T::AccountId {
+		&self.ledger.payee
+	}
+
+	/// Save self to storage.
+	pub(crate) fn save(self) {
+		let key = self.key;
+		self.ledger.save(&key)
+	}
+
+	/// Save self and remove if no delegation left.
+	///
+	/// Returns:
+	/// - true if agent killed.
+	/// - error if the delegate is in an unexpected state.
+	pub(crate) fn save_or_kill(self) -> Result<bool, DispatchError> {
+		let key = self.key;
+		// see if delegate can be killed
+		if self.ledger.total_delegated == Zero::zero() {
+			ensure!(
+				self.ledger.unclaimed_withdrawals == Zero::zero() &&
+					self.ledger.pending_slash == Zero::zero(),
+				Error::<T>::BadState
+			);
+			<Agents<T>>::remove(key);
+			return Ok(true)
+		}
+		self.ledger.save(&key);
+		Ok(false)
+	}
+
+	/// Reloads self from storage.
+	pub(crate) fn refresh(&self) -> Result<Agent<T>, DispatchError> {
+		Self::from(&self.key)
+	}
```

Comment:
now _this_ is actually a function that I would expect it to consume self, because it is possibly dangerous for the call site to have two copies of this value.